### PR TITLE
Remove orderer and add dispatcher

### LINKED ISSUE #1841: Mark messages ready
Enqueue messages to buffer (Scheduler.Submit(id)) after booking.
Mark messages ready (Scheduler.Ready(id)), when their parents are scheduled or confirmed.

### PR DESCRIPTION
**Pull request by capossele**
# Description of change

Fixes #1841, #1849 and #1839

- Remove orderer
- Add dispatcher that uses a workerPool to dispatch scheduled messages to gossip and TSA
- Make use of Ready flag in the scheduler to schedule messages as described in #1841

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

**Comment by capossele** [comment]
Fixes #1841 and #1849

File: packages/tangle/scheduler.go
Comment (piotrm50) [inline packages/tangle/scheduler.go:141]: Check if the message has been scheduled already to avoid iterating through approvers twice - after scheduling and confirmation.

**Review by piotrm50 — CHANGES_REQUESTED** [review]

**Review by piotrm50 — APPROVED** [review]

**Review by piotrm50 — APPROVED** [review]

File: packages/tangle/scheduler.go
Comment (GalRogozinski) [inline packages/tangle/scheduler.go:146]: A good rule on a PR that should be merged is to remove all TODOs

if this is supposed to be done as part of another issue maybe do the "ISSUE xxx:" comment

File: packages/tangle/scheduler.go
Comment (GalRogozinski) [inline packages/tangle/scheduler.go:274]: ditto

File: packages/tangle/scheduler.go
Comment (GalRogozinski) [inline packages/tangle/scheduler.go:309]: This is the only place where we delete from the map correct?
Is it possible to have a memory overflow attack? 

For example, let say someone sends unsolid messages to the node and overflows the map?
Maybe there are also other conditions in which the map won't be pruned...

Btw when we delete from a map in go, it will actually evict the memory? I recall we had some issues with that...

**Review by GalRogozinski — COMMENTED** [review]
Looks nice

Just answer my comments so I can give an approval :-)

File: packages/tangle/scheduler.go
Comment (piotrm50) [reply packages/tangle/scheduler.go:132]: AFAIK those are already logged when they're thrown. However, I removed most of those in a separate PR, as currently the only possible error will be `schedulerutils.ErrInsufficientMana` and it will be logged immediately.

File: packages/tangle/scheduler.go
Comment (piotrm50) [reply packages/tangle/scheduler.go:146]: Good point. The idea was to merge everything and then remove the scheduler bypassed flag to avoid potential conflicts if we did that before merging everything to `dev/drop-head`

File: packages/tangle/dispatcher.go
Comment (capossele) [reply packages/tangle/dispatcher.go:12]: That's a good question! Looking at what we did in the gossip queue, we have 1000 there, so maybe we should do the same here? wdyt?

File: packages/tangle/scheduler.go
Comment (capossele) [reply packages/tangle/scheduler.go:310]: Unsolid messages would not reach the scheduler. This map should always be very small as messages are in general processed already in order. However, if the map should get big, we might want to re-think to use an iteration over the approver instead of using this reverse-map

Comment (capossele) [reply packages/tangle/scheduler.go]: Anyway, I've reverted to use an iteration over the approvers as it seems from the failing tests that the parentsMap was buggy

File: packages/tangle/dispatcher.go
Comment (GalRogozinski) [reply packages/tangle/dispatcher.go:12]: Hmm first I need to ask why we define a size in the first place...
I am guessing the size in the gossip queue achieves 2 purposes:
1. Rate limiter - if messages arrive faster than they are processed the workerpool blocks
2. Saves us from memory overflow

But we also have rate setting mechanism as a function of mana correct? Is it implemented yet?
So the size of the queue just have to be large enough to accomodate that rate

Now looking at the code I understand that all the work in the gossip queue is to trigger an event and you have several go routines that do that for some reason...

The dispatcher is farhter down the pipeline and we don't want it to jam it. So it makes sense it is a bit bigger than the gossip queue. Maybe I would decrease the number of workers of the gossip message worker, but not so important for the prototype I guess...

File: packages/tangle/scheduler.go
Comment (GalRogozinski) [inline packages/tangle/scheduler.go:307]: maybe the approver has one parent confirmed and the other not confirmed? This is enough gor try ready?

File: packages/tangle/scheduler.go
Comment (capossele) [reply packages/tangle/scheduler.go:148]: If a message has been confirmed and not scheduled yet but about to be marked as ready, it should just be unsubmitted as it would be filter out from the dispatcher anyway

File: packages/tangle/scheduler.go
Comment (capossele) [reply packages/tangle/scheduler.go:307]: tryReady calls isReady that checks ALL of its parents

**Review by GalRogozinski — APPROVED** [review]

**Review by piotrm50 — APPROVED** [review]

**Comment by karimodm** [comment]
This fixes #1839

**Pull request merged**
